datumaro/components/dataset.py

```diff
@@ -1248,8 +1249,10 @@ def import_from(
             dataset = cls.from_extractors(*extractors, env=env, merge_policy=merge_policy)
             if eager:
                 dataset.init_cache()
-        except _ImportFail as e:
-            raise e.__cause__
+        except (_ImportFail, DatasetImportError, NotADirectoryError, FileNotFoundError) as e:
+            cause = e.__cause__ if isinstance(e, _ImportFail) else e
+            cause.__traceback__ = e.__traceback__
+            raise DatasetImportError(f"Failed to import dataset '{format}' at '{path}'.") from cause
```

Comment:
Is it better to catch every exception type here rather than specifying some types (L1252)? and why do we need to pull the stack level up one step in L1253-1254? My point is that it can be enough just to make exception chaining on those lines as follows.

```python
except Exception as cause:
    raise DatasetImportError(f"Failed to import dataset '{format}' at '{path}'.") from cause
```

Reply:
We can catch all exceptions like you suggested but I wanted to make it extra clear that Datumaro cares about following exceptions only and other exceptions not caught should be handled by each data format carefully to give informative exception message to users by explicitly rasing a `DatasetImportError`-ish exception.

`_ImportFail` raised simply like this: `raise _ImportFail from exception`. Without it, it shows uninformative additional stack traces.

Current behaviour:
```
Traceback (most recent call last):
  File "/home/andy/workspace/datumaro/datumaro/components/dataset.py", line 1229, in import_from
    env.make_extractor(src_conf.format, src_conf.url, **extractor_kwargs)
  File "/home/andy/workspace/datumaro/datumaro/components/environment.py", line 244, in make_extractor
    return self.extractors.get(name)(*args, **kwargs)
  File "/home/andy/workspace/datumaro/datumaro/plugins/data_formats/coco/base.py", line 531, in __init__
    super().__init__(path, **kwargs)
  File "/home/andy/workspace/datumaro/datumaro/plugins/data_formats/coco/base.py", line 125, in __init__
    self._ctx.error_policy.fail(DatasetImportError("a"))
  File "/home/andy/workspace/datumaro/datumaro/components/importer.py", line 72, in fail
    raise _ImportFail from error
datumaro.components.errors.DatasetImportError: a

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "test2.py", line 20, in <module>
    dataset = Dataset.import_from(
  File "/home/andy/workspace/datumaro/datumaro/components/dataset.py", line 1257, in import_from
    raise DatasetImportError(f"Failed to import dataset '{format}' at '{path}'.") from cause
datumaro.components.errors.DatasetImportError: Failed to import dataset 'coco' at '/mnt/data/dataset/coco_minitrain_25k'.
```

Without it:
```
datumaro.components.errors.DatasetImportError: a

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/andy/workspace/datumaro/datumaro/components/dataset.py", line 1229, in import_from
    env.make_extractor(src_conf.format, src_conf.url, **extractor_kwargs)
  File "/home/andy/workspace/datumaro/datumaro/components/environment.py", line 244, in make_extractor
    return self.extractors.get(name)(*args, **kwargs)
  File "/home/andy/workspace/datumaro/datumaro/plugins/data_formats/coco/base.py", line 531, in __init__
    super().__init__(path, **kwargs)
  File "/home/andy/workspace/datumaro/datumaro/plugins/data_formats/coco/base.py", line 125, in __init__
    self._ctx.error_policy.fail(DatasetImportError("a"))
  File "/home/andy/workspace/datumaro/datumaro/components/importer.py", line 72, in fail
    raise _ImportFail from error
datumaro.components.importer._ImportFail

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "test2.py", line 20, in <module>
    dataset = Dataset.import_from(
  File "/home/andy/workspace/datumaro/datumaro/components/dataset.py", line 1257, in import_from
    raise DatasetImportError(f"Failed to import dataset '{format}' at '{path}'.") from cause
datumaro.components.errors.DatasetImportError: Failed to import dataset 'coco' at '/mnt/data/dataset/coco_minitrain_25k'.
```

Reply:
I definitely agree that L1252 should raise `DatasetImportError()`. However, to make it clear that Datumaro fails to import the dataset, all exceptions should be catched in L1252. This is because there is a possibility that some exception can bypass L1252 and users will not know that this is caused by the dataset import error.

Secondly, is *Without it:* is the one who doesn't have L1253-1254? It looks good to me since 
```console
datumaro.components.errors.DatasetImportError: a <- root cause

The above exception was the direct cause of the following exception: <- chain 1
...
The above exception was the direct cause of the following exception: <- chain 2
...
```

Reply:
1. The exceptions not caught by L1252 should give taceback which users **can** know something went from importing process. It just does not give additional information saying `"Failed to import dataset '{format}' at '{path}'.`
2. I was not saying this is wrong but we can remove uninformative stack `datumaro.components.importer._ImportFail`

Reply:
This one looks much better. Thanks for the suggestion. Fixed in https://github.com/openvinotoolkit/datumaro/pull/992/commits/5fc23c799cf593e304bdf161c7b3d084e6da6e73